libgalois/include/galois/analytics/sssp/sssp.h

```diff
@@ -134,412 +134,32 @@ using SsspEdgeWeight = galois::PODProperty<Weight>;
-                    sdata + graph->template GetEdgeData<EdgeWeight>(e);
-                auto dest = graph->GetEdgeDest(e);
-                auto& ddata = graph->template GetData<NodeDistance>(dest);
-                galois::atomicMin(ddata, new_dist);
-              }
-            }
-          },
-          galois::steal(), galois::loopname("Update"));
-
-    } while (changed.reduce());
-
-    galois::ReportStatSingle("SSSP-Topo", "rounds", rounds);
-  }
-
-  void TopoTileAlgo(Graph* graph, const typename Graph::Node& source) {
-    galois::InsertBag<SrcEdgeTile> tiles;
-
-    graph->template GetData<NodeDistance>(source) = 0;
-
-    galois::do_all(
-        galois::iterate(*graph),
-        [&](const typename Graph::Node& n) {
-          Base::PushEdgeTiles(
-              tiles, graph, n, SrcEdgeTileMaker{n, kDistanceInfinity});
-        },
-        galois::steal(), galois::loopname("MakeTiles"));
-
-    galois::GReduceLogicalOr changed;
-    size_t rounds = 0;
-
-    do {
-      ++rounds;
-      changed.reset();
-
-      galois::do_all(
-          galois::iterate(tiles),
-          [&](SrcEdgeTile& t) {
-            const auto& sdata = graph->template GetData<NodeDistance>(t.src);
-
-            if (t.dist > sdata) {
-              t.dist = sdata;
-              changed.update(true);
-
-              for (auto e = t.beg; e != t.end; ++e) {
-                const Weight new_dist =
-                    sdata + graph->template GetEdgeData<EdgeWeight>(e);
-                auto dest = graph->GetEdgeDest(e);
-                auto& ddata = graph->template GetData<NodeDistance>(dest);
-                galois::atomicMin(ddata, new_dist);
-              }
-            }
-          },
-          galois::steal(), galois::loopname("Update"));
-
-    } while (changed.reduce());
-
-    galois::ReportStatSingle("SSSP-Topo", "rounds", rounds);
-  }
-
-public:
-  galois::Result<void> SSSP(Graph& graph, size_t start_node, SsspPlan plan) {
-    if (start_node >= graph.size()) {
-      return galois::ErrorCode::InvalidArgument;
-    }
-
-    auto it = graph.begin();
-    std::advance(it, start_node);
-    typename Graph::Node source = *it;
-
-    size_t approxNodeData = graph.size() * 64;
-    galois::Prealloc(1, approxNodeData);
-
-    galois::do_all(
-        galois::iterate(graph), [&graph](const typename Graph::Node& n) {
-          graph.template GetData<NodeDistance>(n) = kDistanceInfinity;
-        });
-
-    graph.template GetData<NodeDistance>(source) = 0;
-
-    galois::StatTimer execTime("SSSP");
-    execTime.start();
+GALOIS_EXPORT Result<bool> SsspValidate(
+    graphs::PropertyFileGraph* pfg, size_t start_node,
+    const std::string& edge_weight_property_name,
+    const std::string& output_property_name);
 
-    if (plan.algorithm() == SsspPlan::kAutomatic) {
-      plan = SsspPlan::Automatic(&graph.GetPropertyFileGraph());
-    }
+struct GALOIS_EXPORT SsspStatistics {
+  /// The maximum distance across all nodes.
```

Comment:
Let's add "source node" similar to bfs here as well.

Reply:
I intentionally left it out because it cannot be computed from the graph since multiple nodes may have 0 distance. I think it would have to be passed to Compute which makes it questionably useful as a "statistic". In BFS, it can be computed as the sole 0 in the graph.

Reply:
We could consider attaching the source node as metadata on the property itself. Similar to the [field metadata in Arrow tables](https://arrow.apache.org/docs/cpp/api/datatype.html#_CPPv45fieldNSt6stringENSt10shared_ptrI8DataTypeEEbNSt10shared_ptrIK16KeyValueMetadataEE). If we actually used Arrow Tables under the hood we could just _use_ their metadata system directly. But with our current system we will have to roll our own.
With metadata we could store information about the algorithm configuration which generated the data. The more I think about this, the more I like it. It's kinda of a provenance. We could also store the weight property used for SSSP so that could be looked up later.

Reply:
Yeah, makes sense. Good point.